Fix serialization of font-stretch in font shorthand

darinadler opened the pull request:
#### eecfb48ef6335c3c81b21371d76251336991ccda
<pre>
Fix serialization of font-stretch in font shorthand
<a href="https://bugs.webkit.org/show_bug.cgi?id=246573">https://bugs.webkit.org/show_bug.cgi?id=246573</a>
rdar://problem/101217780

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-font-stretch-expected.txt:
Expect PASS instead of FAIL.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-font-stretch.html:
Expand test to cover handling of keywords, including the normal keyword that should be omitted from
the serialization of the font shorthand. Also check the fontStretch values as well to make the
coverage a little more thorough, and added a case of a value that would round to one of the keywords
to make sure the mapping from percentage to keyword is precise.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-valid-expected.txt:
Update results that had the normal keyword serialized for font-stretch as part of the font
shorthand. There are lots of other problems of the same type here that need to be fixed, so this
still has mostly failing tests. Needs to be revisited.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::fontStretchKeyword): Added. Helper for the function below.
(WebCore::StyleProperties::fontValue const): Added code to treat font-stretch with a
percentage as a special case, and map it to a keyword, even if it wasn&apos;t specified with one.
Also make sure we don&apos;t include &quot;normal&quot; since that&apos;s the font-stretch default.

Canonical link: <a href="https://commits.webkit.org/255628@main">https://commits.webkit.org/255628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8a2e6a59e44704014a784d20601c1690a7a2f4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23747 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102806 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2307 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30627 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85489 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98924 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98764 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1593 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79573 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28513 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71622 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37020 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17155 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34832 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18389 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3901 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38703 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40930 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40633 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37595 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
         return getValueNameAtomString(shorthand.value());
 
+    // Font stretch values can only be serialized as keywords.
+    // While it's not clear the CSS font specification calls for this, web platform tests expect us to convert percentages to keywords.
```
The CSS font spec says percentages for `font-stretch` should get rejected from the `font` shorthand, so that translation is necessary in order to retain a correct value. https://drafts.csswg.org/css-fonts-4/#font-prop . So before this patch, we were effectively serializing to a non valid value.

It is in accordance of step 2 of the CSSOM spec too: https://drafts.csswg.org/cssom/#serializing-css-values

I'd just remove that specific line of the comment re. the spec.

darinadler replied:
Sure, it’s obvious that the old serialization was incorrect. But we could serialize to an empty string in that case. Actually _translating_ percentages to keywords is unusual and unprecedented. And this is what the test expects and requires.

darinadler replied:
I studied the CSSOM specification and I saw **nothing** about converting one type of value (a percentage) into another (a keyword). I’m not sure what you are referring to when you say it’s there. There’s stuff about using a shorter form, and only serializing things that will round trip, but turning a number into a keyword would not be implied by what is written there and i think this is an example of implementers going beyond what is specified.

nt1m replied:
Actually, I would say it's implied by step 1:

1. If this algorithm is invoked with a [list](https://infra.spec.whatwg.org/#list) list:
  1. Let shorthand be the first shorthand property, in [preferred order](https://drafts.csswg.org/cssom/#concept-shorthands-preferred-order), that exactly maps to all of the longhand properties in list.
  2. If there is no such shorthand or shorthand cannot exactly represent the values of all the properties in list, return the empty string.
  3. Otherwise, [serialize a CSS value](https://drafts.csswg.org/cssom/#serialize-a-css-value) from a hypothetical declaration of the property shorthand with its value representing the combined values of the declarations in list.

It would go all the way through step 1.3, and given percentages cannot be represented as they are in the shorthand, they have to be translated. Condition 1.2 is not met here, since the shorthand can exactly represent all the values with some translation.

(I think I mentioned 2. because that step mentions backwards compatibility, but I realize 1. is more accurate)

darinadler replied:
OK, I’ll stop arguing this point with you, and remove the comment. But for the record, please note that step 1.3 says "return the empty string". It does not mention translating anything.

darinadler replied:
It all depends if you think "represent the values" means matching the computed values or matching the specified values. A specified keyword is not the same as a specified percentage in CSS even if they result in the same computed value.

litherum replied:
I don’t know if this is relevant, but this is the opposite behavior from font-weight. I had to write a patch to stop converting from numbers to keywords for font-weight.

darinadler replied:
For now I will land this code, and I guess I don’t need to put that comment in, although for clarity I would like to. The discussion of whether this behavior that Firefox and Chrome did and put into the WPT test is correct or not should probably be a _specification_ discussion, not a WebKit bug report discussion. One good thing about WPT is that if we do change our minds we won’t forget to change the code — it will be a test failure!

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
 
+    // Font stretch values can only be serialized as keywords.
+    // While it's not clear the CSS font specification calls for this, web platform tests expect us to convert percentages to keywords.
+    CSSValueID fontStretchPercentageAsKeyword = CSSValueInvalid;
```
nit: I'd use `auto` here.

darinadler replied:
Will do.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
+    // Font stretch values can only be serialized as keywords.
+    // While it's not clear the CSS font specification calls for this, web platform tests expect us to convert percentages to keywords.
+    CSSValueID fontStretchPercentageAsKeyword = CSSValueInvalid;
+    if (int fontStretchPropertyIndex = findPropertyIndex(CSSPropertyFontStretch); fontStretchPropertyIndex != -1) {
```
Is this a good place to use `notFound` instead of `-1`? (not sure what our guideline is here, especially since this is an `int` and not `size_t`)

darinadler replied:
The `findPropertyIndex` function uses `-1`, and not `notFound`. It’s private to just this source file, and we can change it to use `std::optional` at some point to clean things up. For now, leaving this code as is.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
+            FontSelectionValue valueAsFontSelectionValue { static_cast<float>(valueAsDouble) };
+            auto valueAfterRoundTrip = valueAsFontSelectionValue;
+            if (valueAsDouble != valueAfterRoundTrip)
+                return emptyString();
```
question: What transformation by FontSelectionValue can cause this to be hit?

darinadler replied:
Rounding to the values representable in the fixed point integer. I think one example would be 50.001%. I could add a test showing that if you like.

nt1m replied:
It would be nice to add a comment describing that, test would be nice, but optional.

darinadler replied:
Added a comment and a test.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
+        if (fontStretchPercentageAsKeyword != CSSValueNormal)
+            result.append(result.isEmpty() ? "" : " ", getValueName(fontStretchPercentageAsKeyword));
+        commonValue = { };
+    } else
+        appendFontLonghandValueIfExplicit(CSSPropertyFontStretch, result, commonValue);
```
Took me a bit to get that this is hit in the case of a keyword being set explicitly.

I think we should also omit `CSSValueNormal` in this case too if we can.

This is in-line with the CSSOM spec: https://drafts.csswg.org/cssom/#serializing-css-values

> If component values can be omitted or replaced with a shorter representation without changing the meaning of the value, omit/replace them.

darinadler replied:
Yes, good point. I think I need add a test case for that, because while you caught it the web tests did not! Which is annoying since I always find it slightly confusing and inconvenient to upstream WPT improvements. Maybe you can help me with that part?

The other reason I left that code around is that it handles the “common string” case which might be needed to do the right thing for special values like “inherit” and I might break them if I tried to stop using the helper function. So I decided to limit the change to cases I knew didn’t run into that.

nt1m replied:
> I think I need add a test case for that, because while you caught it the web tests did not! Which is annoying since I always find it slightly confusing and inconvenient to upstream WPT improvements. Maybe you can help me with that part?

Yes! You can probably just extend the current test then use: `export-w3c-test-changes -g HEAD -b 246573 -c --no-linter` (given HEAD is this WebKit commit). It should just be 2 extra lines in the test:

```
    div.style.fontStretch = "normal";
    assert_equals(div.style.font, "medium serif", "The keyword normal should be omitted");
```

You can also make the PR manually from the web UI too: https://github.com/web-platform-tests/wpt/blob/master/css/css-fonts/font-shorthand-serialization-font-stretch.html

I can help with the export part if you don't have time, I'm ok with just making the change in the PR directly in our in-tree files for now.

darinadler replied:
I tried the export but got stuck wondering where I put my OAuth token; I know that seems like a trivial reason to fail at this!

darinadler replied:
https://github.com/darinadler/wpt/tree/wpt-export-for-webkit-246573

nt1m replied:
@darinadler Looks great thanks! can you open a PR on the repo by clicking this link: https://github.com/darinadler/wpt/compare/web-platform-tests:wpt:master...wpt-export-for-webkit-246573?expand=1 and "Create pull request"? This way I can approve it and merge it.

This is basically the step that requires the OAuth token when done from the script, but it can be done manually from the web UI.

nt1m reviewed (approved):
LGTM, mostly nits/questions, thanks!

nt1m reviewed (approved):
LGTM, thanks.

darinadler committed:
Fix serialization of font-stretch in font shorthand
https://bugs.webkit.org/show_bug.cgi?id=246573
rdar://problem/101217780

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-font-stretch-expected.txt:
Expect PASS instead of FAIL.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-shorthand-serialization-font-stretch.html:
Expand test to cover handling of keywords, including the normal keyword that should be omitted from
the serialization of the font shorthand. Also check the fontStretch values as well to make the
coverage a little more thorough, and added a case of a value that would round to one of the keywords
to make sure the mapping from percentage to keyword is precise.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-valid-expected.txt:
Update results that had the normal keyword serialized for font-stretch as part of the font
shorthand. There are lots of other problems of the same type here that need to be fixed, so this
still has mostly failing tests. Needs to be revisited.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::fontStretchKeyword): Added. Helper for the function below.
(WebCore::StyleProperties::fontValue const): Added code to treat font-stretch with a
percentage as a special case, and map it to a keyword, even if it wasn't specified with one.
Also make sure we don't include "normal" since that's the font-stretch default.

Canonical link: https://commits.webkit.org/255628@main
